src/mongocxx/hint.hpp

```diff
@@ -0,0 +1,92 @@
+// Copyright 2015 MongoDB Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#pragma once
+
+#include <mongocxx/config/prelude.hpp>
+
+#include <bsoncxx/builder/stream/document.hpp>
+#include <bsoncxx/builder/stream/helpers.hpp>
+#include <bsoncxx/builder/stream/key_context.hpp>
+#include <bsoncxx/stdx/optional.hpp>
+#include <mongocxx/stdx.hpp>
+
+namespace mongocxx {
+MONGOCXX_INLINE_NAMESPACE_BEGIN
+
+///
+/// Class representing a hint to be passed to a database operation.
+///
+class MONGOCXX_API hint {
+   public:
+    ///
+    /// Constructs a new hint.
+    ///
+    /// @param index
+    ///   Document representing the index to be used.
+    ///
+    explicit hint(bsoncxx::document::value index);
```

Comment:
let's offer both for now until we have some kind of type that represents either a view or value. Since this constructor could be called in hot code in a user's application, it makes sense to offer both the zero-copy ownership transfer (for perf) and the copying constructor (for convenience).